### ISSUE TITLE
Should subtyping be used when declaring modules?

### ISSUE THREAD

**Issue by alexcrichton** [WebAssembly/module-linking#9]
Currently I believe that subtyping primarily comes into account when type-checking `instantiate` statements, ensuring that each item provided is a subtype of the desired type listed in the module type.

Similarly, though, should modules like this be allowed?

```wasm
(module 
  (type (module))
  (module (type 0)
    (func (export ""))
  )
)
```

That is, for modules declared locally should the module type they're specified with only be required to be a subtype of the actual module type itself?

**Comment by rossberg** [comment]
That's a good question, which I also was wondering about. My conclusion was that it's not useful for anything, so there is no reason to allow it. That is, the forward-declared type and the actual type of a nested module should be equivalent, not subtypes.

**Comment by lukewagner** [comment]
Resolved #23.
